airflow/operators/python.py

```diff
@@ -250,25 +251,35 @@ def execute(self, context: Context) -> Any:
         if condition:
             self.log.info("Proceeding with downstream tasks...")
             return condition
-
-        downstream_tasks = context["task"].get_flat_relatives(upstream=False)
+        downstream_tasks = [
+            x for x in context["task"].get_flat_relatives(upstream=False) if not x.is_teardown
+        ]
         self.log.debug("Downstream task IDs %s", downstream_tasks)
 
         if downstream_tasks:
             dag_run = context["dag_run"]
             execution_date = dag_run.execution_date
-
+            if TYPE_CHECKING:
+                assert isinstance(execution_date, DateTime)
             if self.ignore_downstream_trigger_rules is True:
                 self.log.info("Skipping all downstream tasks...")
-                self.skip(dag_run, execution_date, downstream_tasks, map_index=context["ti"].map_index)
+                self.skip(
+                    dag_run=dag_run,
+                    execution_date=execution_date,
+                    tasks=downstream_tasks,
+                    map_index=context["ti"].map_index,
+                )
             else:
                 self.log.info("Skipping downstream tasks while respecting trigger rules...")
                 # Explicitly setting the state of the direct, downstream task(s) to "skipped" and letting the
                 # Scheduler handle the remaining downstream task(s) appropriately.
+                to_skip = [
+                    x for x in context["task"].get_direct_relatives(upstream=False) if not x.is_teardown
+                ]
```

Comment:
I wonder instead of getting the tasks again (which is surpringly not that trivial), it may make more sense to simply filter from `downstream_tasks`

```python
to_skip = [t for t in get_direct_relatives if self.task_id in t.upstream_task_ids]
```

Reply:
when you say not that trivial isn't it just doing `dag.task_dict[x] for x in t.upstream_task_ids` which is not much different?

Reply:
a different way to optimize would to not create the `downstream_tasks` list at all in the `ignore_downstream_trigger_rules is False` case.  then we could just pass to skip `x for x in context["task"].get_direct_relatives(upstream=False) if not x.is_teardown` as a generator since `skip` method can handle that -- i.e. there's not really a need for selectively calling skip.

but stepping back, i'm not adding any new call to either get_flat_relatives or get_direct_relatives -- both calls were there and i don't increase their likelihood of calling or frequency.  i just happened to move to a variable cus it was handy to debug something.

Reply:
OK @uranusjr i went with that optimization... PTAL

Reply:
note: i removed a debug log which emitted the full list of short circuited tasks so that way we could use generator but if you think that's no bueno LMK.

Reply:
ok pushed the optimization / refactor a little further here.  hopefully i don't break something. lemme know what you think.  restored the debug logging, and now it will debug log the actual lists in either case